Allow resetting sharing permissions

nfmohit opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #5445 

## Relevant technical choices

This PR adds a reset sharing permissions action to the Dashboard Sharing settings modal that resets sharing settings to default.

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

tofumatt commented on assets/js/components/dashboard-sharing/DashboardSharingDialog/ResetSharingSettings.js:
```diff
@@ -0,0 +1,99 @@
+/**
+ * DashboardSharingSettingsDialog ResetSharingSettings component.
```
```suggestion
 * Dashboard Sharing "Reset Sharing" component.
```

tofumatt commented on assets/js/components/dashboard-sharing/DashboardSharingDialog/SharingSettings.js:
```diff
@@ -0,0 +1,154 @@
+					closeDialog={ closeDialog }
+					openResetDialog={ openResetDialog }
+				/>
+			</DialogFooter>{ ' ' }
```
What is the whitespace here needed for?

```suggestion
			</DialogFooter>
```

tofumatt commented on assets/js/components/dashboard-sharing/DashboardSharingSettings/Footer.js:
```diff
@@ -79,25 +85,45 @@ export default function Footer( { closeDialog } ) {
-				</Button>
+				<div className="googlesitekit-dashboard-sharing-settings__footer-actions-left">
+					<Link
+						onClick={ () => {
+							openResetDialog();
+						} }
```
This is probably fine here:

```suggestion
						onClick={ openResetDialog }
```

tofumatt reviewed (changes requested):
Maybe this still wasn't clear from the ACs, but this comment mentions not having a modal-on-modal approach: https://github.com/google/site-kit-wp/issues/5445#issuecomment-1190555113

But this PR opens a new modal when clicking on the "Reset sharing permissions" button:

![CleanShot 2022-09-06 at 19 29 15](https://user-images.githubusercontent.com/90871/188712060-b293c081-c7a8-41d5-88b1-c4b7ae044c32.gif)

I think especially on mobile it's a bit awkward of a transition:

![CleanShot 2022-09-06 at 19 35 56](https://user-images.githubusercontent.com/90871/188712970-6200a8d2-8028-4aad-b384-4f2314db4e60.gif)

There are some other UX oddities, like hitting `ESC` closing _both_ modals/screens which might be unexpected. It also looks like it is persisting/saving the settings as soon as the "Reset permissions" button is clicked, which shouldn't happen:

![CleanShot 2022-09-06 at 19 33 08](https://user-images.githubusercontent.com/90871/188712505-c7540e3f-0c38-4a86-abf2-07cef0e0822d.gif)

----

Instead, ideally, the _existing_ Dashboard Sharing modal's content should be replaced with the prompt. And when opening the new prompt, settings shouldn't be saved no matter what.

tofumatt commented:
That looks a lot better, especially on mobile!

What happens when you type `ESC` or click outside the modal? Ideally it should return you to the previous screen, especially on mobile, but I suppose it's okay if it doesn't. Mainly it needs not to save the changes made.

nfmohit commented:
Thank you, @tofumatt! I have addressed all the suggestions. The `ESC` action is also fixed and should return you to the previous screen and does not save the changes made.

tofumatt commented on assets/js/components/dashboard-sharing/DashboardSharingDialog/Footer/index.js:
```diff
@@ -0,0 +1,178 @@
+	const { setValue } = useDispatch( CORE_UI );
+	const { resetSharingSettings } = useDispatch( CORE_MODULES );
+
+	const onApply = useCallback( async () => {
+		setErrorNotice( null );
+
+		const { error } = await saveSharingSettings();
+		if ( error ) {
+			setErrorNotice( error.message );
+			return;
+		}
+
+		trackEvent( `${ viewContext }_sharing`, 'settings_confirm' );
+
+		// Reset the state to enable modules in when not editing or saving.
+		setValue( EDITING_USER_ROLE_SELECT_SLUG_KEY, undefined );
+
+		closeDialog();
+	}, [ viewContext, saveSharingSettings, setValue, closeDialog ] );
+
+	const onReset = useCallback( async () => {
+		setErrorNotice( null );
+		setIsResetting( true );
+
+		const { error } = await resetSharingSettings();
+		if ( error ) {
+			setErrorNotice( error.message );
+			return;
+		}
+
+		setIsResetting( false );
+
+		closeDialog();
+	}, [ closeDialog, resetSharingSettings ] );
+
+	const onCancel = useCallback( () => {
+		trackEvent( `${ viewContext }_sharing`, 'settings_cancel' );
+		closeDialog();
+	}, [ closeDialog, viewContext ] );
```
Rather than calling the variables `onApply`, `onCancel`, etc., it might be nice to name them based on the action they perform. `onCancel` is really `closeResetDialog`, etc.

nfmohit replied:
Understood, thank you!

tofumatt reviewed (approved):
Very nice, this works great with keyboard and mouse, mobile and desktop. Thanks a bunch for all your work on this one 👍🏻

Just a few comments really but I think it's just one tweak and then we can merge this 😃

nfmohit commented on assets/js/components/dashboard-sharing/DashboardSharingDialog/Footer/index.js:
```diff
@@ -0,0 +1,178 @@
+
+	const { saveSharingSettings } = useDispatch( CORE_MODULES );
+	const { setValue } = useDispatch( CORE_UI );
+	const { resetSharingSettings } = useDispatch( CORE_MODULES );
```
Sorry for overlooking this, thank you for fixing!
